src/main/java/com/hivemq/cli/commands/options/TlsOptions.java

```diff
@@ -138,85 +137,298 @@ private boolean useBuiltSslConfig() {
+            keyManagerFactory = null;
         }
+        return keyManagerFactory;
+    }
 
-        try {
-            final Collection<X509Certificate> defaultServerCertificate =
-                    defaultCLIProperties.getServerCertificateChain();
-            if (defaultServerCertificate != null) {
-                if (serverCertificateChain == null) {
-                    serverCertificateChain = new ArrayList<>();
+    private @NotNull KeyManagerFactory createKeyManagerFactoryFromKeystore(
+            final @NotNull Path clientKeystore,
+            final @Nullable String clientKeystorePassword,
+            final @Nullable String clientKeystorePrivateKeyPassword)
+            throws NoSuchAlgorithmException, KeyStoreException, IOException, CertificateException,
+            UnrecoverableKeyException {
+        final KeyManagerFactory keyManagerFactory =
+                KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        if (TlsUtil.isPKCS12Keystore(clientKeystore)) {
+            final KeyStore keyStore = KeyStore.getInstance("PKCS12");
+            final char[] keystorePassword;
+            if (clientKeystorePassword != null) {
+                keystorePassword = clientKeystorePassword.toCharArray();
+            } else {
+                keystorePassword = PasswordUtils.readPassword("Enter keystore password:");
+            }
+            try (final InputStream inputStream = Files.newInputStream(clientKeystore, StandardOpenOption.READ)) {
+                keyStore.load(inputStream, keystorePassword);
+            }
+            keyManagerFactory.init(keyStore, keystorePassword);
+        } else if (TlsUtil.isJKSKeystore(clientKeystore)) {
+            final KeyStore keyStore = KeyStore.getInstance("JKS");
+            final char[] keystorePassword;
+            if (clientKeystorePassword != null) {
+                keystorePassword = clientKeystorePassword.toCharArray();
+            } else {
+                keystorePassword = PasswordUtils.readPassword("Enter keystore password:");
+            }
+            try (final InputStream inputStream = Files.newInputStream(clientKeystore, StandardOpenOption.READ)) {
+                keyStore.load(inputStream, keystorePassword);
+            }
+
+            boolean isDifferentPrivateKeyPassword = false;
+            try {
+                keyManagerFactory.init(keyStore, keystorePassword);
+            } catch (final UnrecoverableKeyException differentPrivateKeyPassword) {
+                isDifferentPrivateKeyPassword = true;
+            }
+            if (isDifferentPrivateKeyPassword) {
+                final char[] keystorePrivateKeyPassword;
+                if (clientKeystorePrivateKeyPassword != null) {
+                    keystorePrivateKeyPassword = clientKeystorePrivateKeyPassword.toCharArray();
+                } else {
+                    keystorePrivateKeyPassword = PasswordUtils.readPassword("Enter keystore private key password:");
```

Comment:
```suggestion
                    keystorePrivateKeyPassword = PasswordUtils.readPassword("Enter private key password:");
```

Reply:
Usually we only call it `privateKeyPassword`.

Reply:
I would not merge the two parameters into one. Adding keystore creates a better readability/ordering on the help page and helps self documentation if the cli is used inside a script.